File: src/evaluator/evaluator.cpp
/* Copyright 2019-2021 Canaan Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <chrono>
#include <nncase/ir/evaluator.h>
#include <nncase/targets/target.h>

using namespace nncase;
using namespace nncase::ir;
using namespace nncase::schedule;
using namespace nncase::runtime;

evaluator::evaluator(const schedule::model_schedule_result &sched)
    : model_eval_(sched)
{
}

void evaluator::enable_ptq(target &target, ir::calibrate_method calib_method)
{
    return model_eval_.enable_ptq(target, calib_method);
}

void evaluator::evaluate(bool before_quant, size_t stage, bool record_output_buffers)
{
    return model_eval_.evaluate(before_quant, stage, record_output_buffers);
}

quantizer *evaluator::quantizer(const module_type_t &module_type)
{
    return model_eval_.module(module_type).quantizer();
}

void evaluator::begin_collect_distribution()
{
    model_eval_.end_sample();
}

void evaluator::end_sample()
{
    model_eval_.end_sample();
}

void evaluator::end_collect_distribution(const std::function<void(size_t cnt, size_t total)> &progress)
{
    model_eval_.end_collect_distribution(progress);
}

evaluate_tensor evaluator::memory_at(const output_connector &conn)
{
    return model_eval_.memory_at(conn);
}

evaluate_tensor evaluator::memory_at(const input_connector &conn)
{
    return model_eval_.memory_at(conn);
}

evaluate_tensor evaluator::input_at(size_t index)
{
    return model_eval_.input_at(index);
}

evaluate_tensor evaluator::output_at(size_t index)
{
    return model_eval_.output_at(index);
}

File: tests/accuracy_test/evaluator.py
# Copyright 2019-2021 Canaan Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# pylint: disable=invalid-name, unused-argument, import-outside-toplevel

from typing import List, Dict, Union, Tuple
import os
import nncase
import numpy as np
import test_utils
import preprocess_utils
from test_utils import *


class Evaluator:
    def run_evaluator(self, compiler, dump_dir):
        evaluator = compiler.create_evaluator(3)

        # transform input
        new_inputs = []
        for idx, value in enumerate(self.inputs):
            new_value = self.transform_input(
                value['data'], self.cfg['compile_opt']['input_type'], "infer")
            new_inputs.append(new_value)

        outputs = []
        number = self.cfg['generator']['inputs']['number']
        for n in range(number):
            # set input
            for idx, value in enumerate(self.inputs):
                evaluator.set_input_tensor(idx, nncase.RuntimeTensor.from_numpy(new_inputs[idx][n]))

            # run
            evaluator.run()

            # get output
            output = self.dump_outputs(evaluator, dump_dir)

            outputs.append(output)
        return outputs

    def dump_outputs(self, evaluator, eval_dir):
        results = []
        compile_opt = self.cfg['compile_opt']
        for i in range(evaluator.outputs_size):
            result = evaluator.get_output_tensor(i).to_numpy()
            if compile_opt['preprocess']:
                if(compile_opt['output_layout'] == 'NHWC' and self.model_type in ['caffe', 'onnx']):
                    result = np.transpose(result, [0, 3, 1, 2])
                elif (compile_opt['output_layout'] == 'NCHW' and self.model_type in ['tflite']):
                    result = np.transpose(result, [0, 2, 3, 1])
                elif compile_opt['output_layout'] not in ["NCHW", "NHWC"] and compile_opt['output_layout'] != "":
                    tmp_perm = [int(idx) for idx in compile_opt['output_layout'].split(",")]
                    result = np.transpose(
                        result, preprocess_utils.get_source_transpose_index(tmp_perm))
            os.makedirs(eval_dir, exist_ok=True)
            if not test_utils.in_ci():
                dump_bin_file(os.path.join(eval_dir, f'nncase_result_{i}.bin'), result)
                dump_txt_file(os.path.join(eval_dir, f'nncase_result_{i}.txt'), result)

            results.append(result)
        return results
